ci_build/view_log.py

```diff
@@ -0,0 +1,91 @@
+import re
+import json
+import os
+import sys
+
+from django.views.generic import TemplateView
+
+from community.views import get_header_and_footer
+from community.git import (get_org_name,
+                           get_owner,
+                           get_deploy_url,
+                           get_upstream_deploy_url)
+
+
+class BuildLogsView(TemplateView):
+    template_name = 'build_logs.html'
+
+    def get_build_info(self):
+        data = {
+            'Org name': get_org_name(),
+            'Owner': get_owner(),
+            'Deploy URL': get_deploy_url(),
+        }
+        try:
+            upstream_deploy_url = get_upstream_deploy_url()
+            data['Upstream deploy URL'] = upstream_deploy_url
+        except RuntimeError:
+            data['Upstream deploy URL'] = 'Not found'
+        return dict(data)
+
+    def get_build_logs(self):
+        log_lines = []
+        log_level_specific_lines = {
+            'INFO': [],
+            'DEBUG': [],
+            'WARNING': [],
+            'ERROR': [],
+            'CRITICAL': []
+        }
+        log_file_path = './_site/community.log'
+        log_file_exists = os.path.isfile(log_file_path)
+        if log_file_exists:
+            with open(log_file_path) as log_file:
+                previous_found_level = None
+                for line in log_file:
+                    log_lines.append(line)
+                    levels = re.findall(r'\[[A-Z]+]', line)
+                    if levels:
+                        level = levels[0]
+                        level = previous_found_level = level[1:-1]
+                        log_level_specific_lines[level].append(line)
+                    elif previous_found_level:
```

Comment:
I am not sure in what condition this line will execute, you have already set  `previous_found_level` to `None`, and I am not seeing getting it overridden anywhere except the `if` condition.

Reply:
It will get execute in case we have a `Traceback` error or a long error message being logged in multiple lines. That log message will be depending upon the previous line which mentions what type of `log` it is!

Reply:
I still did't get you.

Reply:
Allow me to explain this via an example.

Reply:
```bash
2019-06-26 06:50:10,770 [ERROR] django.request: Internal Server Error: /CI/Build/
Traceback (most recent call last):
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/kvgarg/anaconda3/envs/community-env/lib/python3.6/site-packages/django/views/generic/base.py", line 149, in get
    context = self.get_context_data(**kwargs)
  File "/home/kvgarg/GSoC/community/ci_build/view_log.py", line 92, in get_context_data
    context['build_logs_stored'] = list(self.get_build_logs())
TypeError: 'bool' object is not iterable
```

Since the `Traceback` doesn't have any kind of log-level specified, So it is clear that it will depend upon the log-line just before it i.e. `2019-06-26 06:50:10,770 [ERROR] django.request: Internal Server Error: /CI/Build/` therefore the `Traceback` error log will be appended in `Error` dict list.

Reply:
No what I meant is how will you enter into this loop?

```python
something = None
d = 3
if d == 3:
    something = 'something'
elif something:
    print("how can it get printed?")
```
Sorry, if I am missing something.

Reply:
Initially `previous_found_level` is None but once the for-loop begins it **will never** have the value `None`. And, the `levels` will either be empty or containing atmost one entry representing the log-level..
Since, the traceback like log-messages will not contain any type of log-level enclosed in brackets which will return an empty regex match which in turn will execute this code.

Reply:
In the example you have taken the value of `d` Will be getting changed. It will either have a value or will be `None`

Reply:
Understood.

Reply:
Btw, can we break this function into smaller ones?

Reply:
Yeah! In new changes I’ve breaked up this functions into multiple functions